Zoom-to-next-area-animation #419

katia-utochkina opened the pull request:
Ready for review

katia-utochkina committed:
Remove redundant div

katia-utochkina committed:
Add flybtn-holder div

katia-utochkina committed:
Add array with bbox coordinates

katia-utochkina committed:
Create popup

katia-utochkina committed:
Update centroids' minzoom property

katia-utochkina committed:
Delete old mousemove handler

katia-utochkina committed:
Add new mousemove handler

katia-utochkina committed:
Remove old click handler

katia-utochkina committed:
Add click handlder

katia-utochkina committed:
Add a call to flyToNextArea function

katia-utochkina committed:
Add populatePopupOnHover function

katia-utochkina committed:
Add updatePopupOnClick function

katia-utochkina committed:
Add flyToNextArea function

katia-utochkina committed:
Remove redundant CSS

katia-utochkina committed:
Add styling for flybtn

katia-utochkina committed:
Add CSS to flybtn's parent container

katia-utochkina committed:
Add popup styling

katia-utochkina committed:
Fix typo, add var keyword

ramyaragupathy reviewed (commented):
Thanks for working on this @katia-utochkina!

Providing more context for you on this [issue](https://github.com/hotosm/hotosm-website/issues/419):
- HOT undertakes different mapping projects across various countries. You could see all them listed under https://www.hotosm.org/projects/. Like you see in the page, each project has a page dedicated to it on the website.
- A project page provides a high level summary on key statistics, deliverables, people & partners involved and a map viz if applicable. Map view for each project is something that we've introduced recently to showcase the geographical extent of the project. We're still integrating the geographical boundaries for each project in the backend (files are stored in a google drive). So you will see a map/image for projects depending on data availability. Key condition is here a liquid variable `project_extent` in the [front matter](https://jekyllrb.com/docs/front-matter/). If this is empty in a [project file](https://github.com/hotosm/hotosm-website/tree/gh-pages/_projects) then a image replaces a map.
- When a project extent is specified, it links to a project specific [geojson file](http://geojson.org/) in a google drive. Each geojson is a feature collection made up of `n` number of features. Feature here refers to a `polygon` geometry for our case. If it is just one polygon/one feature then it is simple as we highlight just one area. However in reality a project might span across multiple countries leading to > 1 polygon/ 1 feature in the feature collection. 

### Design idea:
- Handling more than one polygon in a project extent is the challenge in your hand. 
- If you look at #419 - @smit1678 has suggested using a button to loop through different polygons. As a first step - you have got the button set right on the page. 
- When a map is loaded, project extent is downloaded from google drive and the entire file is loaded on the map. Though all the polygons are rendered on the map, only the first polygon specified in the  feature collection is highlighted on the map.
- On clicking `Zoom to next area` your code should cycle through the feature collection and position the map to the new polygon bounds. You've got the idea right on flying to a new location with each click. Now your code should pick the geometry dynamically.

### Potential Solution:
Within the [geojson file fetch](https://github.com/hotosm/hotosm-website/blob/gh-pages/js/project-map.js#L137), in the callback chunk, 
- Find the length of feature collection which is an array of features. This gives you the number of polygons you have to deal with.
- If it is just one polygon, then have the map bounds centered on this polygon. There is nothing more to do after this.
- Else if the feature collection length exceeds 1, 
    - initialise a variable- say `currentPolygonIndex` with  value `0`. You could have it defined in global scope and then use it in this block
    - display the `zoom next to button` on map. 
    - Now in your button specific function, increment the global variable by one. While you increment include a condition to see if it is <=(numOfPolygons -1). Else you should reset to `0`. Center the map bounds based on the new global index.

ramyaragupathy commented on js/project-map.js:
```diff
@@ -39,6 +42,11 @@ var map = new mapboxgl.Map({
   style: 'mapbox://styles/hot/cjepk5hhz5o9w2rozqj353ut4'
 })
 
```
Can you please talk more about the design idea for popups?

katia-utochkina committed:
Declare bboxCoordinatesArray

katia-utochkina committed:
Declare currentPolygonIndex and numOfPolygons variables

katia-utochkina committed:
Remove unnecesary code

katia-utochkina committed:
Remove populatePopupOnHover function

katia-utochkina committed:
Alter class names for better readability

katia-utochkina committed:
Remove redundant code

katia-utochkina committed:
Remove redundant code

katia-utochkina committed:
Add a function generating bbox coordinates for each feature

katia-utochkina committed:
Set numOfPolygons equal to features' array's length

katia-utochkina committed:
Add if statement for when to call setMapView function

katia-utochkina committed:
Set currentPolygonIndex to zero

katia-utochkina committed:
Add a call to addFlyBtn and setMapView functions

katia-utochkina committed:
Add flyToNextArea functionality

katia-utochkina committed:
Declare setMapView function

katia-utochkina committed:
Change first argument of fitBounds function

katia-utochkina committed:
Declare addFlyBtn function

katia-utochkina committed:
Remove redundant code

katia-utochkina committed:
Remove redundant CSS

katia-utochkina committed:
Remove redundant CSS

katia-utochkina committed:
Add CSS to change popup's position

katia-utochkina committed:
Hide popup's tip

katia-utochkina committed:
Update names of CSS classes

katia-utochkina commented:
Hi @ramyaragupathy and @smit1678 ,

My PR has been updated.

#### Changes to Review
*  Function to dynamically generate an array of bbox coordinates. Those are used to set the map to the views between which to transition.
*  Functions handling popup display on the page. The original code made the popups stay open and never get closed. This needed to be fixed: if the zoom-to-next area animation was to be added to the map, open popups would have interfered with the user experience. That consideration was why I had to replace the old code handling the removal and display of the popups with the new one. Now, the popups get hidden on the “zoom to next area” button click as well as on the click on the map areas where no markers exist.
*  Modified  functionality allowing users to zoom out of the current location, pan across the map and zoom in on another location.
*  Centroids’ ’minzoom' property. In the original code, centroids do not stay centered on the map and jump around in pan-and-zoom operations. To take care of the wobble, I changed the centroids’ ’minzoom' property from 0 to 10

 @ramyaragupathy, you've asked about the popups' design idea.

 As I’ve understood, there should be a popup attached to each centroid. It should appear at the top left of the screen when a user hovers over a given centroid. Clicking the centroid should expand the popup and display extra information. Another expectation is to have the popup stay open and allow the users to click on a link displayed inside the popup. Finally, clicking  the ‘Zoom to next area’ button should make the popup disappear.

How important is it to have a popup minimised by default and  expanded on click?

So far, not many details get revealed with the  ‘expand on click’ functionality. Why, then, make the users’ interaction with the popup so effortful?  If an extra click is necessary for seeing so little, users can get quickly frustrated. And if that’s the case, their interactions with the page can get much less positive and enjoyable.

If the goal is to create an experience that's positive and something that leads to more interactions, it would probably be helpful to consider that.

Given that, my suggestion would be to turn off the “toggleable/expandable on click” functionality and have the popup display full content by default.

To demonstrate how that would be in practice, I made the suggested changes to the popup behaviour and submitted them as part of my pull request.

katia-utochkina committed:
Change markup for button element

katia-utochkina committed:
Add btn-chevron class to flybtn

katia-utochkina committed:
Add padding to flybtn

katia-utochkina committed:
Remove redundant CSS for flybtn

katia-utochkina committed:
Adjust centroids' minzoom level

ramyaragupathy commented on js/project-map.js:
```diff
@@ -90,59 +100,73 @@ const loadMapLayers = () => {
+
+    var numOfPolygons = projectExtentJSON.features.length;
+
+    if (numOfPolygons.length === 1) {
```
Quick point- `numOfPolygons` already stores the length of the array. So the condition is `numOfPolygons.length` is invalid. Also you could use length only on array variables. Just using `numOfPolygons === 1` should work

ramyaragupathy replied:
I ran through the code with :
```
if (numOfPolygons === 1) {
      setMapView();
}
```
This returns a console error: `camera.js:415 Uncaught TypeError: Cannot read property 'getNorthWest' of undefined`. This is because `currentPolygonIndex` is initialised to zero after this block. This will work when you declare the variable with an intial value of `0`

ramyaragupathy replied:
Fly button should be displayed only if `numOfPolygons` > 1

katia-utochkina replied:
My mistake, sorry.

ramyaragupathy replied:
Great - button displays for just projects with more than one polygons

katia-utochkina committed:
Remove length property from numOfPolygons

katia-utochkina committed:
Rename function

katia-utochkina committed:
Remove unnecessary code

ramyaragupathy reviewed (changes requested):
I still get `camera.js:415 Uncaught TypeError: Cannot read property 'getNorthWest' of undefined` for features with just one polygon. Recheck your condition flow and the variable initialisation

katia-utochkina committed:
Remove unnecessary code

katia-utochkina committed:
Declare index and fitBoundsOptions variables

katia-utochkina committed:
Rename function for better readability

katia-utochkina committed:
Add if statement for when to display button and add button-specific functionality

katia-utochkina committed:
Rename function

katia-utochkina committed:
Declare setMapView function

katia-utochkina committed:
Declare flyToNextArea function

katia-utochkina committed:
Add missing semicolons

ramyaragupathy reviewed (changes requested):
Hey @katia-utochkina - Now I don't see any console errors for projects with greater than one coverage areas. However there is a problem with legend styling across the website on `our-work` page, `country-page` and `project-page`.  There's no sufficient margin around the text and img. Can you please relook at your `base.scss` file changes? Attaching screenshots from your branch:

![image](https://user-images.githubusercontent.com/12103383/56073711-23f9ec00-5dc6-11e9-965e-01774fa6b253.png)

![image](https://user-images.githubusercontent.com/12103383/56073765-ad112300-5dc6-11e9-85cf-0ea37b149714.png)


![image](https://user-images.githubusercontent.com/12103383/56073704-13e20c80-5dc6-11e9-9270-0f141574f9a0.png)

![image](https://user-images.githubusercontent.com/12103383/56073701-0c226800-5dc6-11e9-8a50-22856db9d3d7.png)

katia-utochkina committed:
Fix legend's padding

katia-utochkina commented:
Hi @ramyaragupathy , did you mean legend's padding?

ramyaragupathy commented:
Yes - the spacing around legend icons and text. Also you should bring back `hover details` class  styling in css since this affects the info bar that comes in the top left of `our-work` page and `country-page` maps. This bar is deprived of style now.

![image](https://user-images.githubusercontent.com/12103383/56077152-5e7b7d00-5df6-11e9-8655-9ccc74caa284.png)

katia-utochkina committed:
Add missing CSS for infobar

katia-utochkina commented:
Hi @ramyaragupathy, the info bar's styling has been fixed.

katia-utochkina committed:
Make flybtn centered at map bottom for smaller screens

katia-utochkina committed:
Adjust size, position of popups on map for smaller screens

katia-utochkina committed:
Remove white space from right side of the map

katia-utochkina committed:
Center item in last row of statistics section

katia-utochkina committed:
Add invisible class to hr-v div

katia-utochkina committed:
Remove overlapping line separators in statistics section

katia-utochkina commented:
Hi @ramyaragupathy and @smit1678 ,

I’ve updated my PR. The changes include:

*  Adjusting the placement of the ‘Zoom to next area’ button on smaller screens. Now, on on all screens less than 768px, the button is centered at the bottom of the map. 

*  Changing the popups’ position. That’s because on screens smaller than 768px, they appeared behind the map legend.

<img src="https://user-images.githubusercontent.com/16920088/56248327-3e76e280-60b0-11e9-89e0-0a52dd074e5e.jpg" width="45%"></img> <img src="https://user-images.githubusercontent.com/16920088/56248345-50588580-60b0-11e9-95e6-a84a83e30f70.jpg" width="45%"></img> 

*  Removing the excess white space on the right side of the map  on smaller screen sizes

<img src="https://user-images.githubusercontent.com/16920088/56260989-0092b200-60e2-11e9-9d61-abfa19a2586a.jpg" width="90%"></img> 

* Fixing 2 CSS issues in the home-stats' section:
  * centering the item in the last row of the section;
<img src="https://user-images.githubusercontent.com/16920088/56251013-8d297a00-60ba-11e9-9174-c904a4cfaedb.jpg" width="45%"></img> <img src="https://user-images.githubusercontent.com/16920088/56251002-73883280-60ba-11e9-9507-36e71ca3533f.jpg" width="45%"></img> 

  * removing overlapping separator lines;
<img src="https://user-images.githubusercontent.com/16920088/56251045-b64a0a80-60ba-11e9-975b-533c41a3b83c.jpg" width="45%"></img> <img src="https://user-images.githubusercontent.com/16920088/56251065-c5c95380-60ba-11e9-9e3b-683d80cd0d81.jpg" width="45%"></img>

ramyaragupathy reviewed (changes requested):
@katia-utochkina - 

- Observing a few placement mismatch on country pages legend & pop up.
![image](https://user-images.githubusercontent.com/12103383/56884618-f7adc300-6a87-11e9-8673-16f645f6678d.png)
- Great to know about your modifications for smaller screens 💯  Noticing words spilling out of the black header block in the pop up like the screenshot here.
![image](https://user-images.githubusercontent.com/12103383/56885193-79522080-6a89-11e9-8118-f4c86e713de7.png)
- Also noticing that hover interaction is not working on smaller screens. This is not from your code, seems like an existing issue. https://github.com/hotosm/hotosm-website/issues/543

katia-utochkina committed:
Fix popup placement

katia-utochkina committed:
Fix popup width

katia-utochkina commented:
Hi @ramyaragupathy, the PR has been updated.

ramyaragupathy commented on _layouts/project-item.html:
```diff
@@ -35,20 +37,23 @@ <h1>{{ page.title }}</h1>
 
 <div class="project-map-area">
-    <div id="hover-details" style="display:block" class="hide center-hover-map-area"></div>
     <div id='project-legend-details' style="margin-right: 10px;display:flex" class="center-hover-map-area">
```
Let's move the inline styling to base file

ramyaragupathy commented:
- Legend missing a left margin on projects-page and country-page. Would be nice to use the original `left:5px` here without introducing any legend clash on smaller screens.
- Let's avoid inline styling as much as possible and keep all our styling in base.scss file

katia-utochkina committed:
Fix popup placement

katia-utochkina committed:
Fix legend placement

katia-utochkina commented:
Hi @ramyaragupathy, I've adjusted the popup's and legend's placement. I haven't used any inline CSS in my code. I haven't been able to find it in _layouts/project-item.html
